### PR TITLE
fix(es/parser): Make TS InstantiationExpr parsing more permissive

### PR DESCRIPTION
**Pull request by g-plane**
**Description:**

This PR allows:

```ts
const x4 = f<true>
if (true) {}
```

Previously it's an error.

References:

- https://github.com/microsoft/TypeScript/issues/48654
- https://github.com/microsoft/TypeScript/pull/48659

### CODE REVIEW

File: crates/swc_ecma_parser/src/parser/expr.rs [hunk 1]
```diff
@@ -1904,4 +1904,21 @@ impl<'a, I: Tokens> Parser<I> {
             }
         }
     }
+
+    fn is_start_of_left_hand_side_expr(&mut self) -> PResult<bool> {
+        Ok(is_one_of!(
+            self, "this", "super", "null", "true", "false", Num, BigInt, Str, '`', '(', '[', '{',
+            "function", "class", "new", '/', "/=", IdentRef
```
Comment (kdy1) [inline crates/swc_ecma_parser/src/parser/expr.rs:1911]: Does `/=` start an expression?

Comment (kdy1) [reply crates/swc_ecma_parser/src/parser/expr.rs:1911]: Also I'm not sure about `/`

**Review by kdy1 — CHANGES_REQUESTED** [review]

File: crates/swc_ecma_parser/src/parser/expr.rs (hunk 1 above)
Comment (g-plane) [reply crates/swc_ecma_parser/src/parser/expr.rs:1911]: > Does `/=` start an expression?

I'm not sure, and I just refer to TypeScript's implementation.

> Also I'm not sure about `/`

It's regular expression, I think. It's also referred from TypeScript.

Comment (kdy1) [reply crates/swc_ecma_parser/src/parser/expr.rs:1911]: For `/`, we use `Regex` instead.
https://docs.rs/swc_ecma_parser/latest/swc_ecma_parser/token/enum.Token.html#variant.Regex

Comment (kdy1) [reply crates/swc_ecma_parser/src/parser/expr.rs:1911]: I think our lexer is much more smarter than the one of tsc.
To lex regex, we need the context of parser.
So there are

Comment (kdy1) [reply crates/swc_ecma_parser/src/parser/expr.rs:1911]: So can you remove `/` and `/=` from the list?

Comment (g-plane) [reply crates/swc_ecma_parser/src/parser/expr.rs:1911]: Sure, I will remove them later.

**Review by kdy1 — COMMENTED** [review]
Thank you!

---

swc-bump:
 - swc_ecma_parser

**Review by kdy1 — APPROVED** [review]

**Pull request merged**